addons/web/static/src/core/py_js/py_date.js

```diff
@@ -544,51 +552,157 @@ export class PyRelativeDelta {
+        } else if (namedArgs.yearday) {
+            yearDay = namedArgs.yearday;
+            if (yearDay > 59) {
+                delta.leapdays = -1;
+            }
+        }
+
+        if (yearDay) {
+            for (let monthIndex = 0; monthIndex < DAYS_IN_YEAR.length; monthIndex++) {
+                if (yearDay <= DAYS_IN_YEAR[monthIndex]) {
+                    delta.month = monthIndex + 1;
+                    if (monthIndex === 0) {
+                        delta.day = yearDay;
+                    } else {
+                        delta.day = yearDay - DAYS_IN_YEAR[monthIndex - 1];
+                    }
+                    const absolute = delta.absolute;
+                    if (!absolute.includes("month")) {
+                        absolute.push("month");
+                    }
+                    if (!absolute.includes("day")) {
+                        absolute.push("day");
+                    }
+                    break;
+                }
+            }
         }
         return delta;
     }
 
     /**
-     * @param {PyDate} date
+     * @param {PyDateTime|PyDate} date
      * @param {PyRelativeDelta} delta
-     * @returns {PyDateTime}
+     * @returns {PyDateTime|PyDate}
      */
     static add(date, delta) {
+        date = delta.applyAbsolute(date);
+        const leapDays = delta.getLeapDays(date);
+
         const s = tmxxx(
-            date.year,
-            date.month,
-            date.day + delta.day,
-            delta.hour,
-            delta.minute,
-            delta.second
+            date.year + delta.year,
+            date.month + delta.month,
+            date.day + delta.day + leapDays,
+            date.hour + delta.hour,
+            date.minute + delta.minute,
+            date.second + delta.second
         );
-        return new PyDateTime(s.year, s.month, s.day, s.hour, s.minute, s.second, 0);
+
+        if (delta.weekday !== null) {
+            const sprime = delta.applyWeekday(s);
+            Object.assign(s, sprime);
+        }
+
+        if (delta.hasTime || date instanceof PyDateTime) {
+            return new PyDateTime(s.year, s.month, s.day, s.hour, s.minute, s.second, 0);
+        } else {
+            return new PyDate(s.year, s.month, s.day);
+        }
     }
 
     /**
-     * @param {PyDate} date
+     * @param {PyDateTime|PyDate} date
      * @param {PyRelativeDelta} delta
-     * @returns {PyDateTime}
+     * @returns {PyDateTime|PyDate}
      */
     static substract(date, delta) {
-        const s = tmxxx(
+        const opposite = new PyRelativeDelta();
+        opposite.leapdays = delta.leapdays;
+        opposite.absolute = delta.absolute;
+        opposite.weekday = delta.weekday;
+        opposite.hasTime = delta.hasTime;
+        for (const key of PERIODS) {
+            if (!opposite.absolute.includes(key)) {
+                opposite[key] = -delta[key];
+            } else {
+                opposite[key] = delta[key];
```

Comment:
nope: take relativedelta(day=15,days=3). you get delta.day = 18 absolute and then opposite.day = 18

Reply:
this means we need to distinguish day and days: from 18, you cannot know how it was constructed: 1+17, 2+16,...